string state
bool can_switch_to_external
